### PR TITLE
Fix memory leakage

### LINKED ISSUE #465: Memory leak via com.applitools.eyes.selenium.positioning.PositionProviderFactory static field
Memory leak leads to OOM.

Please see details from Visual VM and heap dumps below:

Screen Shot 2021-02-04 at 13 15 00

Screen Shot 2021-02-04 at 13 16 17

Screen Shot 2021-02-04 at 14 00 00

Screen Shot 2021-02-04 at 13 59 23

### PR DESCRIPTION
**Pull request by ikalinin1**
closes #465 

The same test set with a fix:

Screen Shot 2021-02-04 at 14 28 49

### CODE REVIEW

**Comment by valfirst** [comment]
@danielputerman could you please take a look?

**Comment by itaibh** [comment]
I'm afraid this fix might result in breaking tests, at least some edge cases.
@ikalinin1 @valfirst Can you share your reproduction code? Perhaps we can find a better fix or at least a workaround.
@danielputerman @talgoldy FYI

**Comment by valfirst** [comment]
hi @itaibh you need to run tests for quite long time or limit memory to some small size to hit the issue faster. The provided memory footprints shows that memory leakage happens in the custom cache. Please let us know if the provided details are not enough clear for you.

**Comment by itaibh** [comment]
@valfirst yes, I need to know how many entries are in that field, I couldn't quite understand from the images @ikalinin1 sent by email. The numbers seems way too high, and as you said, the test suite need to be very large and run a very long time. In my experience running hundreds of tests in parallel may take over an hour (we dealt with such long build times by splitting them, but it used to run for over an hour back in the day) and we never encountered a memory issue.

**Comment by ikalinin1** [comment]
Hi @itaibh  more details you could see in ticket #465, all the memory wasted on the image stored in the class field https://github.com/applitools/eyes.sdk.java3/blob/develop/eyes.sdk.core/src/main/java/com/applitools/eyes/MatchWindowTask.java#L24

image

File: eyes.selenium.java/src/main/java/com/applitools/eyes/selenium/positioning/PositionProviderFactory.java [hunk 1]
```diff
@@ -7,23 +7,12 @@
...
 
-import java.util.HashMap;
-import java.util.Map;
+import org.openqa.selenium.WebElement;
 
 public class PositionProviderFactory {
-
-    private static final Map<WebElement, PositionProvider> positionProviders = new HashMap<>();
-
     public static PositionProvider getPositionProvider(Logger logger, StitchMode stitchMode, IEyesJsExecutor executor, WebElement scrollRootElement) {
```
Comment (talgoldy) [inline eyes.selenium.java/src/main/java/com/applitools/eyes/selenium/positioning/PositionProviderFactory.java:14]: I see the this method has no usages. It means that the `positionProviders` map is always empty.
So it makes the method `tryGetPositionProviderForElement` redundant because it will always return null.
So your fix is okay but I don't understand how it caused OOM. Maybe it was something else?

**Comment by itaibh** [comment]
@ikalinin1 @valfirst We can't find a leak in our code, can you please share a code sample that reproduce it? (even over time, it's fine.)
If we look at a sample code we might find the problem and suggest a best practice for how to do what you'd like.

**Comment by ikalinin1** [comment]
Following the thread, I was able to find usage of this method @talgoldy - we were using it in our code as a fallback mechanism to create a position provider dealing with the issue described here #46 

@itaibh  steps to reproduce is to use this method https://github.com/applitools/eyes.sdk.java3/blob/39b7e01efc70e527be0eaf9aafafe7821605250f/eyes.selenium.java/src/main/java/com/applitools/eyes/selenium/positioning/PositionProviderFactory.java#L14. for position provider creation in the code.

**Comment by talgoldy** [comment]
@ikalinin1 So you used the PositionProviderFactory in your test code in the last two years? And to fix the problem you now use the other `getPositionProvider` function after removing the one you used until now?

**Comment by valfirst** [comment]
@talgoldy The caching mechanism causing memory leakage was introduced about 6 months ago: https://github.com/applitools/eyes.sdk.java3/commit/d5f81beba50935e47ae079b193e8a142fa56e52d#diff-07fa537ac950abe0ed430ae5306e3416f257f61f614506cc24b1d85d89101446

**Comment by talgoldy** [comment]
@valfirst @ikalinin1 I need you to be more clear about this.
The function `getPositionProvider` which you removed in this PR is not used by any other class in our SDK. Customers shouldn't use this method as well because it's internal.
If I understood correctly, you used this function to solve an issue in the past.
Please give me more details I'm trying to help here

**Comment by ikalinin1** [comment]
@talgoldy
1. I don't get your idea about the internal nature of the method, because it has a public access level
2. Yes, we used this method for position provider creation.

**Comment by talgoldy** [comment]
@ikalinin1 Yes most of the classes in our sdk are public but users should use only the api in the documentation.
So just to be sure, now that you removed this function (in this PR), which one you will use? The other function with the same name in the same file?

**Comment by ikalinin1** [comment]
@talgoldy Yes, we still will use custom position provider as it was proposed to solve the #46 
https://github.com/applitools/eyes.sdk.java3/pull/47#issuecomment-374936075
If your documentation refers to the "correct" way of creating custom position providers please share a link with me.

The code deleted in the PR according to comments is not used, and on the other hand - once it is used memory leakage will be a thing.
